operator/controllers/csiscaleoperator_controller.go

```diff
@@ -344,17 +344,62 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	// Synchronizing the resources which change over time.
 	// Resource list:
 	// 1. Cluster configMap
-	// 2. Attacher deployment
-	// 3. Provisioner deployment
-	// 4. Snapshotter deployment
-	// 5. Resizer deployment
-	// 6. Driver daemonset
+	// 2. Optional configMap
+	// 3. Attacher deployment
+	// 4. Provisioner deployment
+	// 5. Snapshotter deployment
+	// 6. Resizer deployment
+	// 7. Driver daemonset
+
+	// Synchronizing optional configMap
+	cm, err := r.getConfigMap(instance, config.EnvVarConfigMap)
+	if err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+	if errors.IsNotFound(err) {
+		cmData = map[string]string{}
+		cmDataCopy = map[string]string{}
+		//this means cm is deleted, so set defaults
+		logger.Info("Optional ConfigMap is not found", "ConfigMap", config.EnvVarConfigMap)
+		// setting default values if values are empty
+		setDefaultDriverEnvValues(cmData)
+		logger.Info("Final optional configmap values ", "when the optional configmap is absent", cmData)
+	} else {
+		isValidationNeeded := false
+		// for the first iteration or if there is change in cm data, then validation is required
+		if len(cmDataCopy) == 0 || !reflect.DeepEqual(cm.Data, cmDataCopy) {
+			isValidationNeeded = true
+		}
+		cmDataCopy = cm.Data
+
+		if isValidationNeeded {
+			if err == nil && len(cm.Data) != 0 {
+				cmData = r.parseConfigMap(instance, cm)
+				logger.Info("Final optional configmap values ", "when the optional configmap is present", cmData)
+
+			} else {
+				cmData = map[string]string{}
+				logger.Info("Optional ConfigMap is either not found or is empty, skipped parsing it", "ConfigMap", config.EnvVarConfigMap)
+				// setting default values if values are empty
+				setDefaultDriverEnvValues(cmData)
+				logger.Info("Final optional configmap values ", "when the optional configmap is absent", cmData)
+			}
+		}
+	}
+
+	// Setting the clusterType and presence of CNSA in the final cmData before driver sync
+	if _, exists := cmData[config.ENVClusterConfigurationType]; !exists {
+		logger.Info("Setting the clusterType and presence of CNSA in final cmData")
+		cmData[config.ENVClusterCNSAPresenceCheck] = clusterTypeData[config.ENVClusterCNSAPresenceCheck]
+	}
+
+	logger.Info("Final optional configmap values", "when the sent to syncer is ", cmData)
 
 	// Synchronizing attacher deployment
 	if err := r.removeDeprecatedStatefulset(instance, config.GetNameForResource(config.CSIControllerAttacher, instance.Name)); err != nil {
 		return ctrl.Result{}, err
 	}
-	csiControllerSyncer := clustersyncer.GetAttacherSyncer(r.Client, r.Scheme, instance, restartedAtKey, restartedAtValue)
+	csiControllerSyncer := clustersyncer.GetAttacherSyncer(r.Client, r.Scheme, instance, restartedAtKey, restartedAtValue, cmData)
```

Comment:
It is better to pass only what is required (string values for sidecar CPU and memory limits) as whole cmData is not required (adding more fields to a function call causes the memory used by operator to increase as function call uses stack memory internally)

Reply:
this change is required for other get syncers calls too